pkg/controller/deployment_rolling.go

```diff
@@ -341,3 +365,161 @@ func (dc *controller) taintNodesBackingMachineSets(MachineSets []*v1alpha1.Machi
 
 	return nil
 }
+
+// annotateNodesBackingMachineSets annotates all nodes backing the machineSets
+func (dc *controller) annotateNodesBackingMachineSets(MachineSets []*v1alpha1.MachineSet, annotations map[string]string) error {
+
+	for _, machineSet := range MachineSets {
+
+		klog.V(3).Infof("Trying to annotate nodes under the MachineSet object %q with %s", machineSet.Name, annotations)
+		selector, err := metav1.LabelSelectorAsSelector(machineSet.Spec.Selector)
+		if err != nil {
+			return err
+		}
+
+		// list all machines to include the machines that don't match the ms`s selector
+		// anymore but has the stale controller ref.
+		// TODO: Do the List and Filter in a single pass, or use an index.
+		filteredMachines, err := dc.machineLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		// NOTE: filteredMachines are pointing to objects from cache - if you need to
+		// modify them, you need to copy it first.
+		filteredMachines, err = dc.claimMachines(machineSet, selector, filteredMachines)
+		if err != nil {
+			return err
+		}
+
+		// Iterate through all machines and place the PreferNoSchedule taint
+		// to avoid scheduling on older machines
+		for _, machine := range filteredMachines {
+			if machine.Status.Node != "" {
+				err = AddOrUpdateAnnotationOnNode(
+					dc.targetCoreClient,
+					machine.Status.Node,
+					annotations,
+				)
+				if err != nil {
+					klog.Warningf("Adding annotation failed for node: %s, %s", machine.Status.Node, err)
+				}
+			}
+		}
+		klog.V(2).Infof("Annotated the nodes backed by MachineSet %q with %s", machineSet.Name, annotations)
+	}
+
+	return nil
+}
+
+func (dc *controller) machineSetsScaledToZero(MachineSets []*v1alpha1.MachineSet) bool {
+	for _, machineSet := range MachineSets {
+		if machineSet.Spec.Replicas != 0 && machineSet.Status.AvailableReplicas != 0 && machineSet.Status.FullyLabeledReplicas != 0 && machineSet.Status.ReadyReplicas != 0 && machineSet.Status.Replicas != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// annotateNodesBackingMachineSets annotates all nodes backing the machineSets
+func (dc *controller) removeAnnotationFromNodesBackingMachineSets(MachineSets []*v1alpha1.MachineSet, annotations map[string]string) error {
```

Comment:
This is a helper function, not actively used yet. Happy to remove, if we don't see it being useful.

Reply:
I think we can remove it if it's not used. We can always look back at the commits and reintroduce it if required.

Reply:
Removed.